syntax = "proto3";

package chat;

option go_package = "grpc-chat/chatpb";

service ChatService {
    // Unary RPC: Login user
    rpc Login(LoginRequest) returns (LoginResponse);

    // Server Streaming RPC: Get Live User List
    rpc GetOnlineUsers(Empty) returns (stream OnlineUsersResponse);

    // Client Streaming RPC: User Typing Indicator
    rpc UserTyping(stream TypingRequest) returns (TypingResponse);

    // Bidirectional Streaming RPC: Real-time chat
    rpc ChatStream(stream MessageRequest) returns (stream MessageResponse);
}

message LoginRequest {
    string username = 1;
}

message LoginResponse {
    bool success = 1;
    string message = 2;
}

message Empty {}

message OnlineUsersResponse {
    repeated string users = 1;
}

message MessageRequest {
    string username = 1;
    string message = 2;
}

message MessageResponse {
    string username = 1;
    string message = 2;
}

message TypingRequest {
    string username = 1;
}

message TypingResponse {
    string message = 1;
}
